### DISCUSSION TITLE
Is possible to build AOT for Android like iOS?

### DISCUSSION THREAD

**Discussion by marcussacana** [airsdk/Adobe-Runtime-Support#3127] [Q&A] [answered]
Recently my team released a app after months of work, but some people copied, modified the name, removed our SDK license and released.
So, We are in the initial builds so a lot of things can be improved, but after this problem we are wanting something like is for iOS in android that would make it hard for they modify the app.

**Comment by ajwfrost** [comment, marked as answer]
I'm wondering about different ways we can try to protect against this kind of thing...

* We're just introducing the idea of encrypted assets - which could even be SWF files - where they can be decrypted via an ActionScript call and then loaded into the runtime, with the encryption key being something that has to be pulled down from a server once you've validated that it's the right application (if that's possible ...?)

One issue there is that as soon as you've decrypted it, the asset is sat in memory, and then someone can use tools to dump the memory, scan for the SWF header, and pull out the information..

I'm just thinking for that "identifying if it's your app", we could get the runtime to add a custom header to all HTTP(s) requests that it made, with a hash of the application ID that we get from the OS.. this should mean a webserver can validate that the request for an asset is coming from a particular application.

* AOT would mean the ActionScript byte code is removed, instead it's just all compiled as a native binary with lots of machine code. Makes it pretty tricky to reverse engineer and/or modify. We did some work on this for Android a while ago and have everything in place for the user code to be compiled ahead of time, but we then need to work on the runtime and bootstrap code so that the correct libraries are generated, loaded and initialised properly.. there wasn't really any documentation on how all this should work, so it's been taking a while, but we are aiming to get this support out there!

This of course still leaves a problem for assets such as images/spritesheets, audio files etc. Other than using the encryption mechanism mentioned above - which isn't foolproof - I don't know how else these could be protected. If we try to classify an image as a 'protected asset' and put in place checks to prevent other people from packaging up this same image in a different app, all they have to do is change one pixel slightly and we'd not be able to easily see that it's the same image.


I'm open to ideas for how to protect some of this other content, but yes I think you're right that AOT would massively help with protecting the source code itself... we can check on that work once we've got the AIR 51.0 version out, and see if we can finish it off. (On a separate note - I was also hoping it would help a lot with start-up times..)

thanks

**Reply by bobaoapae** [reply]
Having assets decrypted at runtime by a ActionScript call don't help much in my opnion, anyone can download any as 3 editor, edit my code to print the key on console and voiala, work's it's done.

The whole process of decryption to provide a minimum level of security need to be done at machine code level, embbeded on the runtime.

Some kind of magic trick that will generate random keys of every asset on the package step and add theses keys to the runtime.

also these keys can't be everytime on the same memory region as this it's trivial to anyone with a minimum knowledge to retrieve.

Need to be used some kind of garbage code to be added every new compilation to the runtime libs and just the runtime know the right place to get the keys at runtime doing some calculations.


I already posted this kind of suggestion on some other post you asked for ideas but i don't get any reply so i don't know if this it's not possible or just inviable

**Reply by bobaoapae** [reply]
This need to apply for the swf it's self, just encryption of the assets will not improve much the security, this of course will change when AOT for android get in place.

**Reply by bobaoapae** [reply]
If harman could add some kind of startupt entry point that can be configured via de descriptor to be a native code call instead of a swf file this will help get all this idea to be done by each developer.

as the developer can write a c++ code that will decrypt some initial swf file using they own logic and the start the application. (each developer having the liberty to write they own entry point with they own encryption logic will avoid a lot of stolen codes, as the person or team will need to develop specif scripts to get the data of each application, without having the ability to use the same tatic on several apps)

**Reply by bobaoapae** [reply]
This it's my ideia, if any of this can help i'm more than happy to try this when arrive

**Reply by bobaoapae** [reply]
https://github.com/airsdk/Adobe-Runtime-Support/issues/494

**Reply by marcussacana** [reply]
I'm glad to know that AOT is a upcomming feature, I will be sure to share the news with my team.
About the suggestions, Our current implementation it works like a AIR launcher that downloaded the main swf encrypted, but was expected, it was easy reversed, they probabbly take our initial swf and executed out of our app to get the raw binary data.

We already let our servers reply only when the `AIR` name is found in the user agent, but it was 'emulated' as well, obvious that would be different if we had a random hash in the UserAgent from the native SDK,  In our tests a rooted phone wasn't able to see the https requests (at least not with obvious tools like HttpCanary and/or Http Toolkit), and that is good, then they probabbly get the entrypoint swf and executed in desktop sdk emulator.

The entrypoint hint is interesing, but maybe something similar may be achived with a unsafe initial swf and a ANE, the ANE should be able to validate the APK signature and some others checks (like some API to get info about the current SDK License), we didn't tried that since is our first contact with AIR SDK and we aren't very familiar with the ANE creation yet, so maybe a friendly feature for a custom entry point is really interesing.

Another problem that I saw in the currently SDK is the license be a simple plain text file, it should be some type assimetric data, would be more one detail for the native code consider if should or not decrypt the content, since in the current state it can even be copied for thirdpatry and use as his own license (in our case who stole our app does not had courage for that, but still possible)

**Reply by ajwfrost** [reply]
Wow, loads of ideas .. sorry if I've missed some of these earlier! But just address some of these:

* Encryption/decryption of the SWF - yes our initial attempt at this had been cracked, which used a key that was generated and 'hidden' with the data. That will always be a weak point, trying to have a standalone package that contains everything required to decrypt it .. hence the new APIs that just encrypt/decrypt using a key that needs to be provided separately. So, obtain the key via a call to an HTTPS server, and then you can decrypt

* But yes, having your server work out (automatically) that it's your SWF is something that's a bit trickier. We could provide hashes of the SWF data or similar, but ultimately you would need to have the same info pre-shared on the server, which implies someone else may be able to mock this up and get the same info to then call into your server to get the key.... but we can think about mechanisms for this..

* > Some kind of magic trick that will generate random keys of every asset on the package step and add theses keys to the runtime.. also these keys can't be everytime on the same memory region as this it's trivial to anyone with a minimum knowledge to retrieve.

* I think I see what you're getting at .. e.g. we encrypt the contents using random keys, and then these keys are embedded back into the C++ library somewhere .. but yes, trying to do that in a way that couldn't be detected by someone who is also able to generate a bunch of apps and see how the runtime file changes each time... could be tricky.

* One thing that should be unique (please correct me if I'm wrong!) is the application identifier string. As far as I know, this is used by Google/Apple to identify an application hence must be globally unique. So if we used that string somehow within the process, it could help to protect things..?

* "some kind of startup entry point that can be configured via de descriptor to be a native code call instead of a swf file this will help" -> which I think could be the ANE option, so that people can choose between C++ and Java. But I don't see a huge difference between this vs a start-up SWF that just calls some custom method to decrypt everything else..?

* "the license be a simple plain text file, .. in the current state it can even be copied for thirdparty and use as his own license" - it's probably not watertight, but it shouldn't be possible for you to copy a license.txt file from one app to another..! but at the end of the day, if the license file isn't there or is invalid, all it means is that you get a splash screen appear. Unless we switched to an "everyone must pay" model, it's tricky to do much more here... 

So, my current thinking is:

- we could do some sort of hash generation when packaging up an application, which means you get an output file from ADT and can use that on your server as a kind of pre-shared key.. and then could trigger a "validate me" request from your AS3 code which would ping a message to your server via HTTPS and your server could then validate the connection as definitely originating from your own application. It may need to be a bit of handshaking rather than just a single call.

- encrypting package assets and then updating the runtime with the decryption keys .. maybe combined with some encoding of the application identifier .. could be useful for ensuring that asset files are not easily stripped. But that doesn't stop someone dumping memory at runtime though.. and making this secure could be a challenge unless it's combined with some mechanism that relies on separate delivery of a key...

First step anyway is probably the AOT thing which will help with the code hiding....!

**Reply by bobaoapae** [reply]
Thanks for the reply.

Well I think has a huge improvement of the entry point can be some kind of ane (c++/Java).

If the entry point it's still a common swf, anyone can edit this very easy to output the result of the ane call and get the fully decrypt data.

If the entry point it's a ane that will eventually call c++ code this get a very little tricky, as the majority of people's stealing air applications don't have a great knowledge of assembly as far I know, at least in my case.

So having entry point directly on ane side will be amazing.

I think can be something like that:

The same way Harman added the ability to embbed resources encrypted the initialwindow should support that too, maybe additional parameter on the SWF tag? Or on the descriptor it's self?

Then this initial swf will be encrypted with someone random key.

Then the ane will be called to first decrypt this swf automatically without any AS 3 call, just doing that remove a very huge weak point. For this work perfectly in my ideia this call should expect any response from the ane, as returning data from ane it's not safe from changes of the Java code also, I'm thinking in this process:

Runtime > ane Java > native function c++ (via android ndk) > this call can receive the bytes or just the path for the encrypted swf > some new api to call passing the decrypt bytes to be loaded as initialWindow. 

I don't know from where this call originally come if it's directly from the c++ runtime or this will be called from the Java side in android.

But if this originaly come directly from the c++ bypassing any bytecode languages (as 3 and Java in this case) things get very difficult to intercept for anyone without a good understanding of assembly.

This of course it's not perfect, but add a huge improvement, as currently anyone with a little research download the apk, download some kind of as 3 editor (new ones can easy edit and save directly without need of recompile everything) and have your fully working application.

It's just a bunch of random ideas, as this it's very specific to android and can cause a lot of confusion but I hope to help in anything.

As a lot of my work it's being stolen every day so it's very sad to work with air on Android and Windows.
